e2e/config/vhd.go

```diff
@@ -0,0 +1,176 @@
+// be the resource ID of a managed image or SIG image version, though for now this will always be a SIG image version.
+type VHDResourceID string
+
+func (id VHDResourceID) Short() string {
+	sep := "Microsoft.Compute/galleries/"
+	str := string(id)
+	if strings.Contains(str, sep) && !strings.HasSuffix(str, sep) {
+		return strings.Split(str, sep)[1]
+	}
+	return str
+}
+
+// This is ugly, but I wanted to achieve multiple things:
+// 1. If buildID specified use the image from the build, skip scenario if not found
+// 2. Fetch resource id only once (multiple scenarios can share the VHD)
+// 3. Call fetch method only if tests is scheduled to run (it takes times)
+func (v *VHD) ResourceID() VHDResourceID {
+	if v.BuildResourceID() != "" {
+		return v.BuildResourceID()
+	}
+	return v.NonBuildResourceID()
+}
+
+func (v *VHD) BuildResourceID() VHDResourceID {
+	v.fetchBuildResourceIDOnce.Do(func() {
+		v.buildResourceID, _ = findLatestImageWithTag(v.ImageID, "buildId", VHDBuildID)
+	})
+	return v.buildResourceID
+}
+
+func (v *VHD) NonBuildResourceID() VHDResourceID {
+	v.fetchResourceIDOnce.Do(func() {
+		if v.resourceID != "" {
+			return
+		}
+		v.resourceID, _ = findLatestImageWithTag(v.ImageID, v.VersionTagName, v.VersionTagValue)
+	})
+	return v.resourceID
+}
+
+var ErrNotFound = fmt.Errorf("not found")
+
+func findLatestImageWithTag(imageID, tagName, tagValue string) (VHDResourceID, error) {
+	ctx, cancel := context.WithTimeout(context.TODO(), time.Minute)
+	defer cancel()
+	image, err := parseImageID(imageID)
+	if err != nil {
+		return "", err
+	}
+
+	cred, err := azidentity.NewDefaultAzureCredential(nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to obtain a credential: %v", err)
+	}
+
+	client, err := armcompute.NewGalleryImageVersionsClient(image.subscriptionID, cred, nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to create a new images client: %v", err)
+	}
+
+	pager := client.NewListByGalleryImagePager(image.resourceGroup, image.galleryName, image.imageName, nil)
+	var latestVersion *armcompute.GalleryImageVersion
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return "", fmt.Errorf("failed to get next page: %v", err)
+		}
+		versions := page.Value
+		for _, version := range versions {
+			tag, ok := version.Tags[tagName]
+			if !ok || tag == nil || *tag != tagValue {
+				continue
+			}
+			if latestVersion == nil || version.Properties.PublishingProfile.PublishedDate.After(*latestVersion.Properties.PublishingProfile.PublishedDate) {
+				latestVersion = version
+			}
+		}
+	}
+	if latestVersion == nil {
+		return "", ErrNotFound
+	}
+	return VHDResourceID(*latestVersion.ID), nil
+}
+
+type imageID struct {
+	subscriptionID string
+	resourceGroup  string
+	galleryName    string
+	imageName      string
+}
+
+func parseImageID(resourceID string) (imageID, error) {
+	pattern := `(?i)^/subscriptions/([^/]+)/resourceGroups/([^/]+)/providers/Microsoft\.Compute/galleries/([^/]+)/images/([^/]+)$`
```

Comment:
regex makes me sad, can we use this instead? https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore/arm#ParseResourceID

Reply:
I tried initially. SIG resource id contains more sections than a typical resource. So post-processing with regexes or splitting by `/` is still required after the function call.